### PR TITLE
WithItems Support

### PR DESCRIPTION
**Pull request by kevinbache**
This PR adds for loop support to the KFP DSL.  

Users instantiate loops like so:
```python
from kfp import dsl

@dsl.pipeline(name='my-pipeline', description='A pipeline with a loop.')
def pipeline(my_pipe_param=10):
    loop_args = [{'a': 1, 'b': 2}, {'a': 10, 'b': 20}]
    with dsl.ParallelFor(loop_args) as item:
        op1 = dsl.ContainerOp(
            name="my-in-cop",
            image="library/bash:4.4.23",
            command=["sh", "-c"],
            arguments=["echo op1 %s %s" % (item.a, my_pipe_param)],
        )
```

They currently support multiple operations within the loop and nested operations.  They don't currently support using the output of another operation as the input to loop.

This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1868)

### CODE REVIEW

**Comment by kevinbache** [comment]
This PR adds for loop support to the KFP DSL.  

Users instantiate loops like so:
```python
from kfp import dsl

@dsl.pipeline(name='my-pipeline', description='A pipeline with multiple a loop.')
def pipeline(my_pipe_param=10):
    loop_args = [{'a': 1, 'b': 2}, {'a': 10, 'b': 20}]
    with dsl.ParallelFor(loop_args) as item:
        op1 = dsl.ContainerOp(
            name="my-in-cop",
            image="library/bash:4.4.23",
            command=["sh", "-c"],
            arguments=["echo op1 %s %s" % (item.a, my_pipe_param)],
        )
```

They currently support multiple operations within the loop and nested operations.  They don't currently support using the output of another operation as the input to loop.

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -677,6 +800,9 @@ def compile(self, pipeline_func, package_path, type_check=True):
       yaml.Dumper.ignore_aliases = lambda *args : True
       yaml_text = yaml.dump(workflow, default_flow_style=False)
 
+      if package_path is None:
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py]: What are the use cases where the YAML text is need?
The `Compiler()._compile` method returns the workflow dict which seems more useful.

File: sdk/python/kfp/dsl/__init__.py [hunk 2]
```diff
@@ -15,13 +15,14 @@
 
 from ._pipeline_param import PipelineParam, match_serialized_pipelineparam
 from ._pipeline import Pipeline, pipeline, get_pipeline_conf
-from ._container_op import ContainerOp, UserContainer, Sidecar
+from ._container_op import ContainerOp, UserContainer, Sidecar, BaseOp
```
Comment (Ark-kun) [inline sdk/python/kfp/dsl/__init__.py]: WHy do you think we should expose BaseOp to the users? I'm not sure they're expected to instantiate it.

File: sdk/python/kfp/dsl/_metadata.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 from abc import ABCMeta, abstractmethod
 from .types import BaseType, _check_valid_type_dict, _instance_to_dict
 
+
```
Comment (Ark-kun) [inline sdk/python/kfp/dsl/_metadata.py]: Nit. It's better to do style changes and refactorings in separate PRs so that the main PR diff is smaller.
One way is to create separate commits, tag them with branches and make PRs:
```
master
|
V
Fixed the style -> fix-style branch -> Fix style PR
|
V
Added feature -> add-feature branch -> Added feature PR
```

GitHub allows seeing the diffs of individual commits.

File: sdk/python/kfp/dsl/_pipeline_param.py [hunk 4]
```diff
@@ -11,11 +11,10 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
```
Comment (Ark-kun) [inline sdk/python/kfp/dsl/_pipeline_param.py]: Does this file only have style changes?

File: sdk/python/tests/compiler/compiler_tests.py [hunk 5]
```diff
@@ -591,6 +591,12 @@ def some_pipeline():
       if container:
         self.assertEqual(template['retryStrategy']['limit'], 5)
 
+  def test_withitem_basic(self):
```
Comment (Ark-kun) [inline sdk/python/tests/compiler/compiler_tests.py]: Is it possible to test the actual behavior of the feature instead of just comparing the YAML text?

It would be great if the tests are not starting to fail when some unrelated part (e.g. pipeline name or metadata) changes.

File: sdk/python/tests/compiler/compiler_withitems_test.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+# UUIDs everytime you run this code.
+# This block should be right below the uuid
+# import.
+rd = random.Random()
+rd.seed(0)
+uuid.uuid4 = lambda: uuid.UUID(int=rd.getrandbits(128))
+# -------------------------------------------
+
+
+# @dsl.pipeline(name='my-pipeline')
```
Comment (Ark-kun) [inline sdk/python/tests/compiler/compiler_withitems_test.py]: JFYI, name parameter is optional.

File: sdk/python/tests/compiler/compiler_withitems_test.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+# def pipeline(my_pipe_param=10):
```
Comment (Ark-kun) [inline sdk/python/tests/compiler/compiler_withitems_test.py]: Is this test failing?

File: sdk/python/tests/compiler/testdata/withitem_basic.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import uuid
+import random
+# -------------------------------------------
+# Remove this block to generate different
+# UUIDs everytime you run this code.
+# This block should be right below the uuid
+# import.
+rd = random.Random()
+rd.seed(0)
+uuid.uuid4 = lambda: uuid.UUID(int=rd.getrandbits(128))
```
Comment (Ark-kun) [inline sdk/python/tests/compiler/testdata/withitem_basic.py]: It's usually better to patch a function in KFP library than patch a function in a system library.

We use this for example to override path generation during the tests.

File: sdk/python/tests/compiler/testdata/withitem_nested.yaml [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+      - arguments:
+          parameters:
+          - name: my-pipe-param
+            value: '{{inputs.parameters.my-pipe-param}}'
+        name: my-out-cop
+        template: my-out-cop
+    inputs:
+      parameters:
+      - name: my-pipe-param
+    name: my-pipeline
```
Comment (Ark-kun) [inline sdk/python/tests/compiler/testdata/withitem_nested.yaml]: Mac OS X' sed won't like this EOF =)

**Comment by kevinbache** [comment]
Lol

File: sdk/python/kfp/compiler/compiler.py (hunk 1 above)
Comment (kevinbache) [reply sdk/python/kfp/compiler/compiler.py]: i was using it for visual debugging while i was developing.  i figured why not leave the option for anyone else who ends up working on the compiler.  what do you guys think?

File: sdk/python/kfp/dsl/__init__.py (hunk 2 above)
Comment (kevinbache) [reply sdk/python/kfp/dsl/__init__.py]: leftover from previous hacking

File: sdk/python/kfp/dsl/_metadata.py (hunk 3 above)
Comment (kevinbache) [reply sdk/python/kfp/dsl/_metadata.py]: sure.

File: sdk/python/kfp/compiler/compiler.py [hunk 10]
```diff
@@ -182,7 +183,86 @@ def _get_condition_params_for_ops_helper(group, current_conditions_params):
...
+      if next_op is None:
+        break
+      to_visit.extend(self._get_subgroups(next_op))
+      if isinstance(next_op, dsl.ParallelFor):
+        op_name_to_op[next_op.name] = next_op
+
+    return op_name_to_op
+
+  def _fill_loop_args(self, new_root):
+    """Traverses through graph, plucking up loop_args vars from ops groups and depositing pointers to them on the
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py]: Can you elaborate on this a bit more?

File: sdk/python/tests/compiler/compiler_tests.py (hunk 5 above)
Comment (kevinbache) [reply sdk/python/tests/compiler/compiler_tests.py]: i agree, it'd be nice to have an e2e test as well, but either way, we'll want to include some unit tests too and this is the pattern the repo uses.

File: sdk/python/tests/compiler/compiler_withitems_test.py (hunk 6 above)
Comment (kevinbache) [reply sdk/python/tests/compiler/compiler_withitems_test.py]: 👍

**Comment by Ark-kun** [comment]
Thank you for this great work!
If this PR finished or still WIP?

File: sdk/python/tests/compiler/testdata/withitem_nested.yaml (hunk 9 above)
Comment (kevinbache) [reply sdk/python/tests/compiler/testdata/withitem_nested.yaml]: lol

File: sdk/python/tests/compiler/testdata/withitem_basic.py (hunk 8 above)
Comment (kevinbache) [reply sdk/python/tests/compiler/testdata/withitem_basic.py]: good idea.  changed.

File: sdk/python/kfp/compiler/compiler.py (hunk 10 above)
Comment (kevinbache) [reply sdk/python/kfp/compiler/compiler.py]: you're right, this is a bit vague.

File: sdk/python/tests/compiler/compiler_withitems_test.py (hunk 7 above)
Comment (kevinbache) [reply sdk/python/tests/compiler/compiler_withitems_test.py]: shouldn't be

File: sdk/python/kfp/dsl/_pipeline_param.py (hunk 4 above)
Comment (kevinbache) [reply sdk/python/kfp/dsl/_pipeline_param.py]: possibly

File: sdk/python/tests/compiler/compiler_tests.py (hunk 5 above)
Comment (Ark-kun) [reply sdk/python/tests/compiler/compiler_tests.py]: Unit tests are important and we should have them. However we can usually do better than just comparing YAML output of the whole pipeline (although, checking for the loop compilation behavior might be non-trivial).
As and example of unit tests that test the feature behavior, see 
[test_init_container](https://github.com/kubeflow/pipelines/blob/cd4490554dc8fc67491b60ad0e4ffabe1e84b872/sdk/python/tests/compiler/compiler_tests.py#L614)
[test_op_transformers](https://github.com/kubeflow/pipelines/blob/cd4490554dc8fc67491b60ad0e4ffabe1e84b872/sdk/python/tests/compiler/compiler_tests.py#L589)
[test_set_display_name](https://github.com/kubeflow/pipelines/blob/cd4490554dc8fc67491b60ad0e4ffabe1e84b872/sdk/python/tests/compiler/compiler_tests.py#L551)

File: sdk/python/tests/compiler/compiler_withitems_test.py (hunk 7 above)
Comment (Ark-kun) [reply sdk/python/tests/compiler/compiler_withitems_test.py]: Should there be commented out code here?

**Comment by Ark-kun** [comment]
/lgtm
Are you planning to add support for loop items coming from output of another task?

File: .gitignore [hunk 11]
```diff
@@ -51,3 +51,7 @@ bazel-*
 
 # VSCode
 .vscode
+
```
Comment (Ark-kun) [inline .gitignore]: Hint: You will need less people to approve your PR if you do not modify files in the root directory =)

**Comment by kevinbache** [comment]
> /lgtm
> Are you planning to add support for loop items coming from output of another task?

Yes, we're planning to add WithParams support too, but I'll drop that in another PR.

File: sdk/python/tests/compiler/compiler_tests.py (hunk 5 above)
Comment (kevinbache) [reply sdk/python/tests/compiler/compiler_tests.py]: I agree, less brittle tests would be nice, though the full YAML comparison is more thorough and that is most of how we currently test the compiler.

**Comment by kevinbache** [comment]
> Thank you for this great work!
> If this PR finished or still WIP?

This part's good for now.

File: sdk/python/kfp/compiler/compiler.py [hunk 12]
```diff
@@ -182,7 +184,86 @@ def _get_condition_params_for_ops_helper(group, current_conditions_params):
...
+      to_visit.extend(self._get_subgroups(next))
+
+      # store loop_args from ops groups
+      if hasattr(next, 'loop_args') and next.loop_args is not None:
+        loop_args_name_to_obj[next.loop_args.name] = next.loop_args
+
+      loop_args_names = get_loop_args_names_for_op(next)
+      next.loop_args_names_to_loop_args = {name: loop_args_name_to_obj[name] for name in loop_args_names}
+
+  def _get_subgroups(self, op):
```
Comment (hongye-sun) [inline sdk/python/kfp/compiler/compiler.py]: _get_subopsandgroups?

File: sdk/python/kfp/compiler/compiler.py [hunk 13]
```diff
@@ -182,7 +184,86 @@ def _get_condition_params_for_ops_helper(group, current_conditions_params):
...
+      if next is None:
+        break
+      to_visit.extend(self._get_subgroups(next))
+
+      # store loop_args from ops groups
+      if hasattr(next, 'loop_args') and next.loop_args is not None:
+        loop_args_name_to_obj[next.loop_args.name] = next.loop_args
+
+      loop_args_names = get_loop_args_names_for_op(next)
+      next.loop_args_names_to_loop_args = {name: loop_args_name_to_obj[name] for name in loop_args_names}
```
Comment (hongye-sun) [inline sdk/python/kfp/compiler/compiler.py]: Where is this used?

File: sdk/python/kfp/compiler/compiler.py [hunk 14]
```diff
@@ -182,7 +184,86 @@ def _get_condition_params_for_ops_helper(group, current_conditions_params):
...
+      next_op = self._get_next_group_or_op(to_visit, already_visited)
+      if next_op is None:
+        break
+      to_visit.extend(self._get_subgroups(next_op))
+      if isinstance(next_op, dsl.ParallelFor):
+        op_name_to_op[next_op.name] = next_op
+
+    return op_name_to_op
+
+  def _fill_loop_args(self, new_root):
```
Comment (hongye-sun) [inline sdk/python/kfp/compiler/compiler.py]: Could you move all the related logic in another module file for better readability? This class has already been too big.

File: sdk/python/kfp/dsl/_for_loop.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
+import re
+from typing import List, Union, Dict, Text, Any, Optional
+
+from kfp import dsl
+from kfp.dsl import _metadata
+
+ItemList = List[Union[int, float, str, Dict[Text, Any]]]
+
+
+class LoopArguments(dsl.PipelineParam):
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: Yup, good call.

File: sdk/python/kfp/dsl/_for_loop.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        return re.match(
+            '%s-[0-9a-f]{32}' % cls._loop_item_placeholder_name,
+            param_name
+        ) is not None
+
+    def __init__(self, items: ItemList, code: Text):
+        super().__init__(name=self.make_name(code))
+
+        if not isinstance(items, (list, tuple)):
+            raise ValueError("Expected list or tuple, got {}.".format(type(items)))
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: Better

File: sdk/python/kfp/dsl/_for_loop.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+        if isinstance(items[0], dict):
+            subvar_names = set(items[0].keys())
+            for item in items:
+                if not set(item.keys()) == subvar_names:
+                    raise ValueError("If you input a list of dicts then all dicts should have the same keys. "
+                                     "Got: {}.".format(items))
+
+            # then this block creates loop_args.variable_a and loop_args.variable_b
+            for subvar_name in subvar_names:
+                setattr(self, subvar_name, dsl.LoopArgumentVariable(self.name, subvar_name))
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: good call.  added a check.

File: sdk/python/kfp/dsl/_for_loop.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+from kfp.dsl import _metadata
+
+ItemList = List[Union[int, float, str, Dict[Text, Any]]]
+
+
+class LoopArguments(dsl.PipelineParam):
+    """Class representing the arguments that are looped over in a ParallelFor loop in the KFP DSL.
+    This doesn't need to be instantiated by the end user, rather it will be automatically created by a
+    ParallelFor ops group."""
+    _loop_item_placeholder_name = 'loop-item-placeholder'
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: yup

File: sdk/python/kfp/dsl/_for_loop.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+from typing import List, Union, Dict, Text, Any, Optional
+
+from kfp import dsl
+from kfp.dsl import _metadata
+
+ItemList = List[Union[int, float, str, Dict[Text, Any]]]
+
+
+class LoopArguments(dsl.PipelineParam):
+    """Class representing the arguments that are looped over in a ParallelFor loop in the KFP DSL.
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: not a bad idea, but I'm going to skip here since that's a lot of extra documentation to keep in sync, there's nothing complicated to keep track of here, and I rarely if ever see that in Python code in the wild.  Good idea if it violates those conditions, though.

File: sdk/python/kfp/dsl/_for_loop.py [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+    _loop_item_placeholder_name = 'loop-item-placeholder'
+
+    @classmethod
+    def param_is_this_type(cls, param: dsl.PipelineParam):
+        """Return True if the given param is a LoopArgument param."""
+        return cls.name_is_loop_arguments(param.name)
+
+    @classmethod
+    def make_name(cls, code: Text):
+        """Make a name for this parameter."""
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: sure

File: sdk/python/kfp/dsl/_for_loop.py [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+    ParallelFor ops group."""
+    _loop_item_placeholder_name = 'loop-item-placeholder'
+
+    @classmethod
+    def param_is_this_type(cls, param: dsl.PipelineParam):
+        """Return True if the given param is a LoopArgument param."""
+        return cls.name_is_loop_arguments(param.name)
+
+    @classmethod
+    def make_name(cls, code: Text):
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: sure

File: sdk/python/kfp/dsl/_for_loop.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+        """Make a name for this parameter."""
+        return '{}-{}'.format(cls._loop_item_placeholder_name, code)
+
+    @classmethod
+    def name_is_loop_arguments(cls, param_name: Text):
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: sure

File: sdk/python/kfp/dsl/_for_loop.py [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+        """Return True if the given parameter name looks like it came from a loop arguments parameter."""
+        return re.match(
+            '%s-[0-9a-f]{32}' % cls._loop_item_placeholder_name,
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: it doesn't and I don't.  I shortened it from 32 characters to 8 just to be sure and tightened up the loop name base too

File: sdk/python/kfp/dsl/_for_loop.py [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class LoopArguments(dsl.PipelineParam):
+    """Class representing the arguments that are looped over in a ParallelFor loop in the KFP DSL.
+    This doesn't need to be instantiated by the end user, rather it will be automatically created by a
+    ParallelFor ops group."""
+    _loop_item_placeholder_name = 'loop-item-placeholder'
+
+    @classmethod
+    def param_is_this_type(cls, param: dsl.PipelineParam):
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: done

File: sdk/python/kfp/dsl/_ops_group.py [hunk 25]
```diff
@@ -141,3 +148,36 @@ def __init__(self, name):
...
+  with dsl.ParallelFor([{'a': 1, 'b': 10}, {'a': 2, 'b': 20}]) as item:
+    op1 = ContainerOp(..., args=['echo {}'.format(item.a)])
+    op2 = ContainerOp(..., args=['echo {}'.format(item.b])
+  ```
+  and op1 would be executed twice, once with args=['echo 1'] and once with args=['echo 2']
+  """
+  TYPE_NAME = 'for_loop'
+
+  @staticmethod
+  def _get_code():
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_ops_group.py]: staticmethod makes it easy to override for tests.

File: sdk/python/kfp/compiler/compiler.py [hunk 26]
```diff
@@ -202,31 +283,41 @@ def _get_inputs_outputs(self, pipeline, root_group, op_groups, opsgroup_groups,
...
           if not op.is_exit_handler:
-            for g in op_groups[op.name]:
-              inputs[g].add((full_name, None))
+            for group_name in op_groups[op.name][::-1]:
+              # if group is for loop group and param is that loop's param, then the param
+              # is created by that for loop ops_group and it shouldn't be an input to
+              # any of its parent groups.
+              inputs[group_name].add((param.full_name, None))
+              if group_name in op_name_to_for_loop_op:
+                # loop_group.loop_args.name = 'loop-item-placeholder-99ca152eb887400e992909bfa0ef1050'
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/compiler/compiler.py]: it's meant as an explanatory comment.  i added more info to clarify.

File: sdk/python/kfp/compiler/compiler.py (hunk 13 above)
Comment (kevinbache) [reply sdk/python/kfp/compiler/compiler.py]: it isn't.  in fact this whole function is out of date; I'll remove it.

File: sdk/python/kfp/compiler/compiler.py (hunk 14 above)
Comment (kevinbache) [reply sdk/python/kfp/compiler/compiler.py]: removing method

File: sdk/python/kfp/compiler/compiler.py (hunk 12 above)
Comment (kevinbache) [reply sdk/python/kfp/compiler/compiler.py]: changed

File: sdk/python/kfp/dsl/_for_loop.py [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+        Args:
+            loop_args_name: the name of the loop args parameter that this LoopArgsVariable is attached to.
+            this_variable_name: the name of this LoopArgumentsVariable subvar.
+
+        Returns: The name of this loop args variable.
+        """
+        return '{}{}{}'.format(loop_args_name, cls.SUBVAR_NAME_DELIMITER, this_variable_name)
+
+    @classmethod
+    def param_is_this_type(cls, param: dsl.PipelineParam):
```
[the remark this answers is not in the document]
Comment (kevinbache) [reply sdk/python/kfp/dsl/_for_loop.py]: sure

File: sdk/python/kfp/dsl/_ops_group.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (hongye-sun) [reply sdk/python/kfp/dsl/_ops_group.py]: Python style guide avoids to use staticmethod. Here's the explanation: http://go/pystyle#decorators

File: sdk/python/tests/compiler/compiler_withitems_test.py [hunk 28]
```diff
@@ -0,0 +1,93 @@
...
+
+    op_out = dsl.ContainerOp(
+        name="my-out-cop",
+        image="library/bash:4.4.23",
+        command=["sh", "-c"],
+        arguments=["echo %s" % my_pipe_param],
+    )
+
+
+# @dsl.pipeline(name='my-pipeline')
```
Comment (hongye-sun) [inline sdk/python/tests/compiler/compiler_withitems_test.py]: remove?

**Comment by Ark-kun** [comment]
`git checkout origin/master .gitignore`
